R/shoji-catalog.R

```diff
@@ -240,28 +240,111 @@ setMethod("emails", "ShojiCatalog", function (x) getIndexSlot(x, "email"))
 #' @export
 as.list.ShojiCatalog <- function (x, ...) lapply(names(index(x)), function (i) x[[i]])
 
-#' Utility to get a more human-readable view of a Shoji Catalog
+#' A utility to return a data.frame from a ShojiCatalog.
 #'
-#' @param x ShojiCatalog or subclass
+#' Some of the attributes of a `ShojiCatalog` will not naturally fit in
+#' a conventional data.frame. For example, an array variable contains a list of
+#' subvariables, and these subvariables will not easily fit in a single row of a
+#' data.frame. In this case, the list of subvariables are stored in a
+#' list-column in the resulting data.frame.
+#'
+#' @param x `ShojiCatalog` or subclass
 #' @param keys character vector of attribute names from each catalog tuple to
 #' include in the result. Default is TRUE, which means all.
-#' @param rownames See [`base::data.frame`], the `row.names`
-#' argument, to which this is passed in `data.frame`. The difference here
-#' is that if `rownames` is explicitly set as `NULL`, the resulting
-#' object will not have row names set. By default, row names will be the URLs
-#' of the catalog tuples.
+#' @param rownames See [base::data.frame()] for the `row.names`
+#' argument. The difference here is that if `rownames` is explicitly set as
+#' `NULL`, the resulting object will not have row names set. By default, row
+#' names will be the URLs of the catalog tuples.
+#' @param list_columns A character vector of the names of the attributes that
+#' should be stored in a list-column.
 #' @param ... additional arguments passed to `data.frame`
 #' @return a `data.frame` view of the catalog
 #' @export
-catalogToDataFrame <- function (x, keys=TRUE, rownames, ...) {
-    default.rownames <- missing(rownames)
-    if (default.rownames) {
-        rownames <- NULL
+catalogToDataFrame <- function (x, keys=TRUE, rownames = NULL,
+    list_columns = c("subvariables", "subvariables_catalog"),
+    ...) {
+
+    index <- lapply(index(x), function(a) a[keys])
+    ## Return an empty dataframe if the function is called on an empty catalog
+    if (length(index) == 0) {
+        message("Catalog is empty, returning an empty dataframe")
+        return(data.frame())
+    } else {
+        ### The following code is equivalent to out <- purrr::map_df(index, entry_to_df)
+        ################
+        entry_list <- lapply(index, entryToDF, list_col_names = list_columns)
+        names   <- unique(unlist( lapply(entry_list, names)))
+        out <- data.frame(matrix(nrow = length(entry_list), ncol = length(names)))
+        names(out) <- names
+
+        for (i in seq_along(entry_list)) {
+            for (j in  names(entry_list[[i]])) {
+                # Overwriting a entry in a dataframe with a longer list
+                # can trigger a warning, which needs to be suppressed.
```

Comment:
What's the warning? I'd read that as a sign that the code isn't doing the right thing.

Reply:
That is is :)